File: PSS_B3.m
% Parameter Subset Selection Algorithm with PCA B3 (PSS_B3.m)

% Determines unidentifiable parameters using 
% CSS Algorithm 4.3 from 'Robust Parameter Identifiability Analysis'
% (Pearce et al. 2022)

% Written by Kate Pearce

function [UnId, Id, c] = PSS_B3(dydq, eta, k)
% Inputs:
    % dydq: sensitivity matrix (double, n x p with n >= p)
    % eta (optional, double): threshold for info matrix rank; default 1e-8
    % k (optional, integer): rank, or number identifiable params   
% Outputs:
    % UnId: unidentifiable parameter indices
    % Id: identifiable parameter indices
    % c: success criteria for rank revealing fact

% initialize values
UnId = [];

p = size(dydq, 2);
Id = 1:p;

c = [];

% do an initial QR decomposition
[Q, R] = qr(dydq, 0); 
[~, SingVals, V] = svd(R, 'econ'); 

if (~exist('k', 'var')) || (isempty(k) == 1)
    if (~exist('eta', 'var')) || (isempty(eta) == 1) || (nargin < 2)
        eta = 1e-8; % default
    end

    % if eta provided: find k
    SingVals = diag(SingVals);
    ind = find((SingVals./SingVals(1)) > eta); 
    k = length(ind); 

end

% if k provided 

if k > 0
    W = V(:,1:k)';
    
    % move column w largest norm to front
    P_tild = 1:p;

    col_norms = zeros(1,p);
    for wcol = 1:p
        col_norms(wcol) = norm(W(:,wcol),2);
    end
    [~, max_ind] = max(col_norms);

    P_tild([1 max_ind]) = P_tild([max_ind 1]);
    
    % compute QR of R*P_tild
    B = R(:,P_tild);
    [Q_tild, R_tild] = qr(B);
    
    Q = Q*Q_tild;
    R = R_tild;
    P = P_tild;
    
    % main loop
    for l = 2:k % safe indexing: won't execute if k <= 1
        
        R_22 = R(l:end,l:end);
        
        [~,~,V_l] = svd(R_22);

        W = V_l(:,1:(k-l+1))';
        col_norms = zeros(1,p-l+1);
        for wcol = 1:(p-l+1)
            col_norms(wcol) = norm(W(:,wcol),2);
        end

        [~, max_ind] = max(col_norms);
        
        P_tild_loop = 1:(p-l+1);
        P_tild_loop([1 max_ind]) = P_tild_loop([max_ind 1]);
        
        % compute qr decomposition of R_22 * P_tild
        [Q_tild_loop, R_tild_loop] = qr(R_22(:,P_tild_loop));
        
        temp = P_tild_loop+ones(size(P_tild_loop))*(l-1); % shift for # iters done
        temp_ind = [1:(l-1),temp];

        % update
        Q = Q*blkdiag(eye(l-1), Q_tild_loop);
        P(1:p) = P(temp_ind);
        R(l:end,l:end) = R_tild_loop;

    end
    
    Id = P(1:k);  
    UnId = P(k+1:end);
    [abs_err, rel_err, cond_S, cond_S1] = SuccessCheck(dydq, P, k, SingVals); % criteria 
    c = [abs_err, rel_err; cond_S, cond_S1] ;
            
end

end

File: PSS_Python/PSS_B3.py
import numpy as np
from scipy.linalg import qr, svd
from PSS_B1 import success_check

def PSS_B3(dydq, eta=1e-8, k=None):
    """
    Parameter Subset Selection Algorithm with PCA B3
    
    Args:
        dydq: Sensitivity matrix (n x p with n >= p)
        eta: Threshold for info matrix rank (default: 1e-8)
        k: Rank, or number of identifiable params
        
    Returns:
        UnId: Unidentifiable parameter indices
        Id: Identifiable parameter indices
        c: Success criteria for rank revealing factorization
    """
    # Initialize values
    UnId = []
    p = dydq.shape[1]
    Id = np.arange(p)
    c = []
    
    # Initial QR decomposition
    Q, R = qr(dydq, mode='economic')
    _, sing_vals, V = svd(R, full_matrices=False)
    
    if k is None:
        # If eta provided: find k
        sing_vals_diag = np.diag(sing_vals) if len(sing_vals.shape) > 1 else sing_vals
        ind = np.where((sing_vals_diag / sing_vals_diag[0]) > eta)[0]
        k = len(ind)
    
    if k > 0:
        W = V[:, :k].T
        
        # Move column with largest norm to front
        P_tild = np.arange(p)
        col_norms = np.linalg.norm(W, axis=0)
        max_ind = np.argmax(col_norms)
        
        P_tild[0], P_tild[max_ind] = P_tild[max_ind], P_tild[0]
        
        # Compute QR of R*P_tild
        B = R[:, P_tild]
        Q_tild, R_tild = qr(B, mode='economic')
        
        Q = Q @ Q_tild
        R = R_tild
        P = P_tild.copy()
        
        # Main loop
        for l in range(1, k):  # l starts from 1 to k-1
            R_22 = R[l:, l:]
            
            _, _, V_l = svd(R_22, full_matrices=False)
            
            W = V_l[:, :(k-l)].T
            col_norms = np.linalg.norm(W, axis=0)
            max_ind = np.argmax(col_norms)
            
            P_tild_loop = np.arange(p-l)
            P_tild_loop[0], P_tild_loop[max_ind] = P_tild_loop[max_ind], P_tild_loop[0]
            
            # Compute QR decomposition of R_22 * P_tild
            Q_tild_loop, R_tild_loop = qr(R_22[:, P_tild_loop], mode='economic')
            
            temp = P_tild_loop + l  # Shift for number of iterations done
            temp_ind = np.concatenate([np.arange(l), temp])
            
            # Update
            Q_block = np.block([[np.eye(l), np.zeros((l, Q_tild_loop.shape[1]))],
                              [np.zeros((Q_tild_loop.shape[0], l)), Q_tild_loop]])
            Q = Q @ Q_block
            P[:] = P[temp_ind]
            R[l:, l:] = R_tild_loop
        
        Id = P[:k]
        UnId = P[k:]
        abs_err, rel_err, cond_S, cond_S1 = success_check(dydq, P, k, sing_vals)
        c = np.array([[abs_err[0], rel_err[0]],
                     [abs_err[1], rel_err[1]],
                     [cond_S, cond_S1]])
    
    return UnId, Id, c